### PR TITLE
Make the library work with SoftWire in addition to TwoWire

### PR DESCRIPTION
**Pull request by ghaug**
Hi Rob,
I have the problem that my project requires three sensors and I only have one HW I2C interface. I therefore added SoftWire support to your library. It works fine but the implementation is quite ugly. SoftWire doesn't inherit from TwoWire and I therefore have to check each time wether it's a SoftWire ore TwoWire interface. I'm not concerned about the performance, this isn't that fast anyway, but still it's not nice. Moreover, it adds a dependency to SoftWire, of course. I therefore would fully understand should you not want this change.
Best, Gunter

### CODE REVIEW

**Comment by RobTillaart** [comment]
@ghaug 

Thanks for the addition, I will have a look at the code later this week (quite busy maintaining almost 200 libs)

In the past I had this very same discussion in another library and I recall these points
- softWire did not inherit from TwoWire (as you said) making the implementation a bit complex.
- maintainability becomes more complex due to extra code paths / checks.
- footprint of the library increased (maybe not in your version)
- there are multiple "soft_I2C" implementations and choosing one is based upon which criteria?

Having thought about the softWire and in my opinion the problem is the solution: inheritance.

Can you write **SHT31_SW** which inherits from **SHT31** and replaces the low level code with SoftWire calls?

**Comment by RobTillaart** [comment]
@ghaug 

Note that the build failed because the softwire library needs to be referenced in  SHT31/.arduino-ci.yml

The example below is from my Adler library, it uses an external library.
Works only if softWire is a registered library IIRC,.

```yml
platforms:
  rpipico:
    board: rp2040:rp2040:rpipico
    package: rp2040:rp2040
    gcc:
      features:
      defines:
        - ARDUINO_ARCH_RP2040
      warnings:
      flags:

packages:
  rp2040:rp2040:
    url: https://github.com/earlephilhower/arduino-pico/releases/download/global/package_rp2040_index.json

compile:
  # Choosing to run compilation tests on 2 different Arduino platforms
  platforms:
    - uno
    # - due
    # - zero
    # - leonardo
    - m4
    - esp32
    - esp8266
    # - mega2560
    - rpipico

  libraries:
    - "printHelpers"
 ```

**Comment by RobTillaart** [comment]
With respect to your problem of using multiple sensors, you could use a hardware multiplexer (?)

- https://github.com/RobTillaart/MAX14661
- https://github.com/RobTillaart/TCA9548

File: library.properties [hunk 1]
```diff
@@ -8,4 +8,4 @@ category=Sensors
 url=https://github.com/RobTillaart/SHT31
 architectures=*
 includes=SHT31.h
-depends=
\ No newline at end of file
+depends=SoftWire
```
Comment (RobTillaart) [inline library.properties:11]: Not agree with this dependency as it is not true for all projects ;)

**Comment by RobTillaart** [comment]
@ghaug 

Had a quick look at the code and it looks not too bad. 
In fact my proposal to create a derived class only need to implement a few functions and you do not need all the if's 

(example of derived class see - https://github.com/RobTillaart/MCP_DAC - if you're not familiar with that)

**Comment by ghaug** [comment]
Hi Rob,
thanks for looking into this and your quick reply and sorry for being late.

> Note that the build failed because the softwire library needs to be referenced in SHT31/.arduino-ci.yml

Sorry about that, I'm completely new to Arduino. I'll fix it.

> Had a quick look at the code and it looks not too bad.
In fact my proposal to create a derived class only need to implement a few (6 ?) functions and you do not need all the if's

I'll look into it, but unfortunately only over the weekend.

Thanks for your patience!

**Comment by RobTillaart** [comment]
SoftWire seems to be a non registered library.
Have to verify that.
Don't know how to include these in build CI
Do you have a link to its repository?

**Comment by ghaug** [comment]
> Do you have a link to its repository?


Yes, it's here: https://github.com/stevemarple/SoftWire

**Comment by ghaug** [comment]
> Can you write SHT31_SW which inherits from SHT31 and replaces the low level code with SoftWire calls?

Yes, in fact I did this to make my project work and I can live with it, no problem. However, it might break if SHT31 changes internally. Or do you mean adding SHT31_SW (or something like that) to your library?

**Comment by ghaug** [comment]
Regarding the failing test:
SoftWire can apparently be installed, at least there is this output:
`Installing dependency of <library.properties>: 'SoftWire'...                   ✓`
However, there is no mention of the respective directory in a -I option, only this:
`-I/opt/hostedtoolcache/Ruby/2.6.10/x64/lib/ruby/gems/2.6.0/gems/arduino_ci-1.5.0/cpp/arduino -I/opt/hostedtoolcache/Ruby/2.6.10/x64/lib/ruby/gems/2.6.0/gems/arduino_ci-1.5.0/cpp/unittest -I/home/runner/Arduino/libraries/SHT31`
Therefore the include fails.
Sorry, if this a dumb question but as I said I'm new to Arduino. How is that supposed to work? Do I have to add the include path to .arduino-ci.yml manually?

**Comment by RobTillaart** [comment]
> do I have to add the include path to .arduino-ci.yml manually?

I have to dive into the documentation, as I did not use it so far.

Easiest way is to get the library registered. 
Then it is also available in the library manager.

Alternative is to make a TwoWire compatible version of SoftWire, that would make it usable for almost any I2C library.

**Comment by RobTillaart** [comment]
Had a look at https://github.com/stevemarple/SoftWire/blob/master/LICENSE.txt == GNU
including it in my library impacts on my (MIT) license

(one of the reasons to write libraries is to have minimal restrictions)

**Comment by ghaug** [comment]
OK, at least I found out what would have to be added to .arduino-ci.yml. However, this is still not enough as util/atomic.h is not mocked in arduino-ci (avr/interrupt.h isn't either, but that's empty anyway). I mocked them locally and now arduino_ci.rb completes successfully for me.

Anyway, I think it's a moot point given the license issues.

**Comment by ghaug** [comment]
I've added a script that adds a temporary mock for util/atomic.h to arduino_ci. This works in my repo and should at make the tests run successfully. I can create a pull request with arduino_ci to add the mock.

**Comment by RobTillaart** [comment]
I copied your .h and .cpp to my develop environment, compiled **SHT31_demo.ino**  and got serious errors.

The changes forces the user to download SoftWire Library even if they do not use it.

```
Alternatives for SoftWire.h: [SoftI2CMaster@2.1.6]
ResolveLibrary(SoftWire.h)
  -> candidates: [SoftI2CMaster@2.1.6]
"C:\\Users\\Rob\\AppData\\Local\\Arduino15\\packages\\arduino\\tools\\avr-gcc\\7.3.0-atmel3.6.1-arduino7/bin/avr-g++" -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -flto -w -x c++ -E -CC -mmcu=atmega328p -DF_CPU=16000000L -DARDUINO=10819 -DARDUINO_AVR_UNO -DARDUINO_ARCH_AVR "-IC:\\Users\\Rob\\AppData\\Local\\Arduino15\\packages\\arduino\\hardware\\avr\\1.8.6\\cores\\arduino" "-C:\\Users\\Rob\\AppData\\Local\\Temp\\arduino_build_971794\\sketch\\SHT31_demo.ino.cpp.o"

.....

In file included from c:\users\rob\appdata\local\arduino15\packages\arduino\tools\avr-gcc\7.3.0-atmel3.6.1-arduino7\avr\include\avr\io.h:99:0,
                 from c:\users\rob\appdata\local\arduino15\packages\arduino\tools\avr-gcc\7.3.0-atmel3.6.1-arduino7\avr\include\avr\pgmspace.h:90,
                 from C:\Users\Rob\AppData\Local\Arduino15\packages\arduino\hardware\avr\1.8.6\cores\arduino/Arduino.h:28,
                 from C:\Users\Rob\AppData\Local\Temp\arduino_build_971794\sketch\SHT31_demo.ino.cpp:1:
D:\Rob\WORK\Arduino\libraries\SoftI2CMaster\src/SoftI2CMaster.h: In function 'void i2c_wait_scl_high()':
D:\Rob\WORK\Arduino\libraries\SoftI2CMaster\src/SoftI2CMaster.h:243:31: error: 'SCL_PORT' was not declared in this scope
 #define SCL_IN  (_SFR_IO_ADDR(SCL_PORT) - 2)
                               ^
D:\Rob\WORK\Arduino\libraries\SoftI2CMaster\src/SoftI2CMaster.h:280:23: note: in expansion of macro 'SCL_IN'
      : : [SCLIN] "I" (SCL_IN), [SCLPIN] "I" (SCL_PIN));
                       ^~~~~~
D:\Rob\WORK\Arduino\libraries\SoftI2CMaster\src/SoftI2CMaster.h:243:31: note: suggested alternative: 'SCL_OUT'
 #define SCL_IN  (_SFR_IO_ADDR(SCL_PORT) - 2)
                               ^
D:\Rob\WORK\Arduino\libraries\SoftI2CMaster\src/SoftI2CMaster.h:280:23: note: in expansion of macro 'SCL_IN'
      : : [SCLIN] "I" (SCL_IN), [SCLPIN] "I" (SCL_PIN));
```

**Comment by RobTillaart** [comment]
installing SoftWire

```Alternatives for AsyncDelay.h: []
ResolveLibrary(AsyncDelay.h)
  -> candidates: []
In file included from D:\Rob\WORK\Arduino\libraries\SHT31/SHT31.h:14:0,
Multiple libraries were found for "SoftWire.h"
                 from D:\Rob\WORK\Arduino\libraries\SHT31\examples\SHT31_demo\SHT31_demo.ino:9:
D:\Rob\WORK\Arduino\libraries\SoftWire\src/SoftWire.h:8:10: fatal error: AsyncDelay.h: No such file or directory
 #include <AsyncDelay.h>
          ^~~~~~~~~~~~~~
compilation terminated.
 Used: D:\Rob\WORK\Arduino\libraries\SoftWire
 Not used: D:\Rob\WORK\Arduino\libraries\SoftI2CMaster
```

**Comment by RobTillaart** [comment]
Installing AsyncDelay ==> build successful

```
Sketch uses 8406 bytes (26%) of program storage space. Maximum is 32256 bytes.
Global variables use 544 bytes (26%) of dynamic memory, leaving 1504 bytes for local variables. Maximum is 2048 bytes.
```

However 0.3.7 gave
```
Sketch uses 6716 bytes (20%) of program storage space. Maximum is 32256 bytes.
Global variables use 542 bytes (26%) of dynamic memory, leaving 1506 bytes for local variables. Maximum is 2048 bytes.
```
The 2 bytes RAM is relative neglectable

However, 1690 bytes extra PROGMEM due to two libraries not used is not acceptable imho. 
With a derived class this would (should) not happen as the base class does not use them.

**Comment by ghaug** [comment]
> With a derived class this would (should) not happen as the base class does not use them.

Yes, I think that's the way to go.

**Comment by RobTillaart** [comment]
I'll notify the maintainer of the build-CI of your script, might be useful for him to learn how it could be done, OK?

**Comment by ghaug** [comment]
> I'll notify the maintainer of the build-CI of your script, might be useful for him to learn how it could be done, OK?

If you want to, sure, but I can create a pull request tomorrow. The script just creates the file in the appropriate location. It's just like all the other mocked headers.

**Comment by ghaug** [comment]
I've moved SoftWire support to the derived class SHT31sw. I had to make the affected functions virtual and the member variables protected in your code. It will add a vtable to instances of SHT31, but I hope that is acceptable.
If you prefer, or this is more in accordance with Arduino-Style we could move the code to separate files.

**Comment by RobTillaart** [comment]
Quite busy so won't have much time to review, sorry

**Comment by ghaug** [comment]
No problem! If you are generally interested in the change, I'll add a unit test and few words in the README, and possibly an example.

**Comment by RobTillaart** [comment]
HI Gunther,

Wrote an experimental wrapper class for **SoftwareWire** that implements the **TwoWire** interface.
As I do not have a test setup ready, can you give it a try?
It is rather minimalistic and therefore should be easy to understand by you.

[TwoWireSW.zip](https://github.com/RobTillaart/SHT31/files/10994987/TwoWireSW.zip)

If it works well it would allow (almost) all my I2C based libraries to work with **SoftwareWire**

**Comment by RobTillaart** [comment]
The demo code with the **TwoWire** wrapper class

```
Sketch uses 8646 bytes (26%) of program storage space. Maximum is 32256 bytes.
Global variables use 546 bytes (26%) of dynamic memory, leaving 1502 bytes for local variables. Maximum is 2048 bytes.
```

230 bytes more than your first version 
(although I may not compare a Wire with a SoftwareWire version, they appear in the same order of magniitude)

Advantage of the wrapper class (which I hope will work)
- sensor class do not need any change.
- usable for many other I2C classes.
- maintenance is only in the wrapper class - not in all sensors classes.
- I have an I2C scanner class that should be able to use it - https://github.com/RobTillaart/I2C_SCANNER
- wrapper class can be a pattern for other software I2C implementations.

**Comment by ghaug** [comment]
Hey Rob,

> As I do not have a test setup ready, can you give it a try?

Will do tomorrow!

> So 140 bytes PROGMEM more than 0.3.7 due to the changes, but that is 10x better than the previous design 👍👍👍
> It also uses 10 bytes more RAM and unclear why (not investigated)

I guess that's the price for the late binding of the three functions. The instances have a vtable now and the compiler needs to insert a little code to get the actual targets out.

**Comment by RobTillaart** [comment]
Thanks for testing, have a busy day tomorrow so it might take a while before I will be online.

**Comment by ghaug** [comment]
> Advantage of the wrapper class (which I hope will work)

> sensor class do not need any change.
> usable for many other I2C classes.
> maintenance is only in the wrapper class - not in all sensors classes.
> I have an I2C scanner class that should be able to use it - https://github.com/RobTillaart/I2C_SCANNER
> wrapper class can be a pattern for other software I2C implementations.

Yes, that's a very good approach! Unfortunately the wrapper doesn't work out of the box. I could make it compile with a few tweaks but it still doesn't work. I'll debug it tomorrow or on Sunday, today I'm busy. Why don't you put a repo on GitHub? Will make it easier to collaborate.

**Comment by RobTillaart** [comment]
Quickly created a repo - https://github.com/RobTillaart/TwoWireSW

**Comment by ghaug** [comment]
Hi Rob,
I just got around to testing the wrapper in https://github.com/RobTillaart/TwoWireSW. Unfortunately it doesn't work and I'm afraid, there is a fundamental problem. The `requestFrom` functions of TwoWire are not virtual so calls to these functions  (e.g. in SHT31, which I used to test) are bound early by gcc. When SHT31.cpp is compiled the only implementation of  `requestFrom` known to gcc is the one in TwoWire and as it is not declared virtual, the compiler inserts a call directly to this implementation.

**Comment by ghaug** [comment]
On the positive side, my PR with arduino_ci (regarding the problem with util/atomic.h) has been merged into the main brach. With the next release of arduino_ci the work-around with the script int this PR will not be required any more.

**Comment by ghaug** [comment]
As I wrote via our other channel, now less changes to your class.

**Comment by RobTillaart** [comment]
@ghaug 
Created SHT31 version 0.3.8 with
- virtual bool isConnected();
- virtual bool writeCmd(uint16_t cmd);
- virtual bool readBytes(uint8_t n, uint8_t *val);

Next step is to fix SHT31_SW as derived class

**Comment by RobTillaart** [comment]
@ghaug 

Fixed the build of https://github.com/RobTillaart/SHT31_SW
Can you please test if it works for you?

If so this PR can be closed, and I will publish the SHT31_SW in the Arduino library registry

**Comment by RobTillaart** [comment]
@ghaug 
Thanks for testing, appreciated!

Really liked the interaction. It was a pity that the TwoWire wrapper did not work as that would bring SoftWire to many (I2C) libraries at once. Still very happy with exploring the different alternatives and the chosen solution. The design can be used for other I2C based libraries too or other I2C implementations (on request only). So time well spend.

There are indeed a few cosmetic things to clean. These might be fixed in a next release.

The patch for "Atomic" will probably cause a failing build after the update of the build-ci. 
I'll make a note to check when updating the SHT31_SW library (not expected soon).

Finally I need to publish the SHT31_SW in the Arduino registry, will do that right away.

Again, thanks for the issue, thinking, writing and testing!
(you may close this PR as it will not be merged)

**Comment by ghaug** [comment]
Thanks to you!

**Pull request closed without merging**